src/python/WMCore/MicroService/MSTransferor/MSTransferor.py

```diff
@@ -676,3 +730,115 @@ def _diskPNNs(self, pnnList):
             else:
                 diskPNNs.add(pnn)
         return diskPNNs
+
+    def updateSites(self, rec):
+        """
+        Update sites API provides asynchronous update of Site info.
+
+        :param rec: JSON payload with the following data structures:
+                    {'workflow': <wflow name>, 'SiteWhiteList' ['T1', ...], 'SiteBlackList': ['T2',...]}
+        :return: acknowledge dict to upstream caller (ReqMgr2)
+        """
+        # preserve provided payload to local file system
+        wflow = rec['workflow']
+        status = self.saveData(wflow, rec)
+        if status == 'ok':
+            return []
+        err = MSTransferorStorageError(rec)
+        self.logger.error(err)
+        return [err.error()]
+
+    def saveData(self, wflow, rec):
+        """
+        Save workflow data to persistent storage
+        :param wflow: name of workflow
+        :param rec: record to save
+        :return: status of this operation
+        """
+        try:
+            fname = '{}/{}'.format(self.storage, wflow)
+            with open(fname, 'w', encoding="utf-8") as ostream:
+                ostream.write(json.dumps(rec))
+            return 'ok'
+        except Exception as exp:
+            msg = "Unable to save workflow '%s' to storage=%s. Error: %s", wflow, self.storage, str(exp)
+            self.logger.exception(msg)
+            return str(exp)
+
+    def readData(self, wflow):
+        """
+        Read workflow data from persistent storage
+        :return: data or None
+        """
+        fname = '{}/{}'.format(self.storage, wflow)
+        data = None
+        with open(fname, 'r', encoding="utf-8") as istream:
+            data = json.load(istream)
+        return data
+
+    def _updateSites(self, wflow, rseList):
```

Comment:
If my memory does not fail me, last discussion was that this method is completely irrelevant. Am I missing anything?

With that said, this is the place we read the workflow information (calling `readData()`). So it has to be moved somewhere else.

Reply:
I left the method itself where I only read data and update workflow. But I removed from its implementation any rucio creation rules. Since implementation is more then one page long I think that it is better to keep it as separate method. It seems to me that you didn't look through the code and only replied based on previous discussion where we agreed to remove replication calls from it. Keeping this method makes more clean code where it is called, otherwise I need to put additional page (from this method codebase) into already quite long `execute` method.

Reply:
What we discussed last week - and that I have been recommending since the beginning of this implementation - is to use everything that is provided by MSTransferor.

I am not able to find where I reported it, but I already mentioned a couple of times that this method is missing much of the functionality that is needed for input data placement, **hence** I suggest to go through the standard `execute()` method instead of crafting something separate.

Some of the issues in this method at this moment are:
* does not convert PSN to PNN/RSE (site names to storage names)
* does not consider RSE quota
* does not consider campaign and pileup configuration

Reply:
I still need further clarifications since the whole logic you refer of using execute is already in place. Putting aside lack of proper conversion from PSN, RSE quote and pileup configuration

Let me iterate:
- the execute method needs to know if workflow requires the update and so far this is done with `_updateSites` method which after reading it from persistent storage set a flag for `dataPlacement`
- this is done before we call `makeTransferRequest` call, in other words:
```
def execute()
...
      # this only load wf from persistent storage and decide to set a dataPlacement flag for wf:
      errors = self._updateSites(wflow, rseList)
      ....
      # perform actual logic of transfer
      success, transfers = self.makeTransferRequest(wflow, rseList)
```
The `makeTransferRequest` is doing properly data movement, either create rules or move rules. Therefore, what you are asking is to put code from `_updateSites` into execute which I argue is more than 1 page long and will explode further very long `execute` logic or keep it under separate method.

Am I missing something?

Reply:
Well, how do we obtain which workflows need a data placement when site lists have changed? The current implementation of `_updateSites` does that. The steps b, c, d in fact are implemented.

Reply:
this thread discussion is resolved via private zoom chat and changing the logic of execute algorithm